### PR TITLE
Add core API page

### PR DESCRIPTION
**Pull request by ianagbip1oti**
An attempt at documenting core api concepts.

### CODE REVIEW

**Comment by CoreyShupe** [comment]
Should we address the events and how they're done?

**Comment by ianagbip1oti** [comment]
I've added the "Event Stream" as a concept in the core api.

After thinking about it Events, Resources, Requests, etc are really just data that is passed around through the api. These could be mentioned in this doc, but then I think it starts to lose a little bit of focus. The focus currently is on the APIs that allow you to do stuff, without getting into specifics of what data you need to pass around and receive.

File: docs/core_api.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+but for which there is no response expected.
+Examples include sending a message, or adding a role.
+
+
+## Repositories
+
+`Repository`s are the source of information from Discord.
+This could include, for example, information about guilds, channels, or users.
+
+`Repository` implmentations may perform caching, gateway requests, or HTTP requests
```
Comment (CoreyShupe) [inline docs/core_api.md]: Spelling error.
See http://www.dictionary.com/browse/implementation
"implmentations"

**Review by CoreyShupe — CHANGES_REQUESTED** [review]

File: docs/core_api.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
+# Core API
+
+The core API of ErisCasper.java is designed to be minimal and powerful.
+
+It exposes the concepts of `Bot`s, the event stream (`Observable<Event>`), `Action`s, and `Repository`s as
+described below.
+
+There also exists the concept `Gateway` and `Routes`.
```
Comment (CoreyShupe) [inline docs/core_api.md]: awk, maybe rephrase, also the plurality of `concept` suggests 1 concept.

File: docs/core_api.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+The return type is a `Completable` as defined by RxJava.
+This Completable will be subscribed to by ErisCasper.java.
+
+`Bot`s can be composed. Multiple implementations of `Bot` may be merged into a single `Bot`.
+
+
+## Event Stream
+
+The event stream is the events sent from Discord.
```
Comment (CoreyShupe) [inline docs/core_api.md]: `consists of the events sent from discord`
A stream would consist of items, not be the array of events. imo it works better. Up for debate.

File: docs/core_api.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+This includes events such as message sends, new guild members, etc.
+
+The event stream uses the `Observable` type from RxJava to make the events available to bot developers.
+
+
+## Actions
+
+`Action`s are interactions with the Discord API that are expected to change its state,
+but for which there is no response expected.
+Examples include sending a message, or adding a role.
```
Comment (CoreyShupe) [inline docs/core_api.md]: A comma isn't necessary after `message` when referring to 2 items of the same context. </nit>

File: docs/core_api.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+
+## Repositories
+
+`Repository`s are the source of information from Discord.
+This could include, for example, information about guilds, channels, or users.
```
Comment (CoreyShupe) [inline docs/core_api.md]: `This could include: information about guilds; channels; users; and many more forms of data which need to be cached`
May work better or some other variation.
Maybe `This could include, for example: information about guilds; channels; or users.`
I see it as better just because we use phrases in the list, `information about guilds`, which doesn't follow the normal convention of having a list of items. Up for debate though.

File: docs/core_api.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+`Repository` implmentations may perform caching, gateway requests, or HTTP requests
+as required.
+
+## Internal Concepts
+
+### Gateway
+
+Provide access to the (Discord Gateway)[https://discordapp.com/developers/docs/topics/gateway].
```
Comment (CoreyShupe) [inline docs/core_api.md]: `provides`, the active form of the word

File: docs/core_api.md [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+This is a websocket provided by Discord used for tasks such as authentication and receiving events.
```
Comment (CoreyShupe) [inline docs/core_api.md]: authentication
sending gateway commands
and receiving events

Has more than just the 2 uses described.

File: docs/core_api.md [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+### Routes
+
+Provides access to Discord's REST API.
```
Comment (CoreyShupe) [inline docs/core_api.md]: We should reference this the same way we do the gateway.
https://discordapp.com/developers/docs/reference < this is the link I'd refer to.

**Review by CoreyShupe — CHANGES_REQUESTED** [review]

File: docs/core_api.md [hunk 9]
```diff
@@ -5,8 +5,8 @@ The core API of ErisCasper.java is designed to be minimal and powerful.
 It exposes the concepts of `Bot`s, the event stream (`Observable<Event>`), `Action`s, and `Repository`s as
 described below.
 
-There also exists the concept `Gateway` and `Routes`.
-These are lower level wrappers around the Discord API and considered internal
+At a lower level are `Gateway`s and `Routes`.
```
Comment (CoreyShupe) [inline docs/core_api.md]: do we have multiple gateways?

Comment (CoreyShupe) [reply docs/core_api.md]: nvm ignore this.

**Review by CoreyShupe — DISMISSED** [review]

**Review by CoreyShupe — APPROVED** [review]
